File: protocol-generator/c/src/wire.c
/*
 * wire.c — message builders (§3.2 EXECUTE / §3.3 EXECUTE_RESPONSE) + error result +
 * empty-params + resource-target. Framing (the 4-byte BE length prefix + body) lives in
 * transport.c; this file builds the envelope-payload entities.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#include "wire.h"

#include <stdlib.h>
#include <string.h>

/* put a text key + value into a map, taking ownership; returns EC_OK / frees both on err. */
static ec_status put_kv(ec_value *map, const char *key, ec_value *val)
{
    if (!val) {
        return EC_ERR_OOM;
    }
    ec_value *k = ec_text(key);
    if (!k) {
        ec_value_free(val);
        return EC_ERR_OOM;
    }
    ec_status st = ec_map_put(map, k, val);
    if (st != EC_OK) {
        ec_value_free(k);
        ec_value_free(val);
    }
    return st;
}

ec_status ec_make_execute(const char *request_id, const char *uri, const char *operation,
                          ec_entity *params, const uint8_t *author, const uint8_t *capability,
                          ec_value *resource, ec_entity **out)
{
    ec_status st = EC_ERR_OOM;
    ec_value *m = ec_map();
    ec_value *params_cbor = NULL;
    if (!m) {
        goto cleanup;
    }
    if ((st = put_kv(m, "request_id", ec_text(request_id))) != EC_OK) { goto cleanup; }
    if ((st = put_kv(m, "uri", ec_text(uri))) != EC_OK) { goto cleanup; }
    if ((st = put_kv(m, "operation", ec_text(operation))) != EC_OK) { goto cleanup; }
    st = ec_entity_to_cbor(params, &params_cbor);
    if (st != EC_OK) { goto cleanup; }
    if ((st = put_kv(m, "params", params_cbor)) != EC_OK) { goto cleanup; }
    params_cbor = NULL;
    if (author) {
        if ((st = put_kv(m, "author", ec_bytes(author, 33))) != EC_OK) { goto cleanup; }
    }
    if (capability) {
        if ((st = put_kv(m, "capability", ec_bytes(capability, 33))) != EC_OK) { goto cleanup; }
    }
    if (resource) {
        if ((st = put_kv(m, "resource", resource)) != EC_OK) { goto cleanup; }
        resource = NULL;
    }
    st = ec_entity_make_owning("system/protocol/execute", m, out);
    return st;          /* m consumed by make_owning */
cleanup:
    ec_value_free(params_cbor);
    ec_value_free(resource);
    ec_value_free(m);
    return st;
}

ec_status ec_make_response(const char *request_id, uint64_t status, ec_entity *result,
                           ec_entity **out)
{
    ec_status st = EC_ERR_OOM;
    ec_value *m = ec_map();
    ec_value *result_cbor = NULL;
    if (!m) {
        goto cleanup;
    }
    if ((st = put_kv(m, "request_id", ec_text(request_id))) != EC_OK) { goto cleanup; }
    if ((st = put_kv(m, "status", ec_int_u(status))) != EC_OK) { goto cleanup; }
    st = ec_entity_to_cbor(result, &result_cbor);
    if (st != EC_OK) { goto cleanup; }
    if ((st = put_kv(m, "result", result_cbor)) != EC_OK) { goto cleanup; }
    result_cbor = NULL;
    return ec_entity_make_owning("system/protocol/execute/response", m, out);
cleanup:
    ec_value_free(result_cbor);
    ec_value_free(m);
    return st;
}

ec_status ec_error_result(const char *code, const char *message, ec_entity **out)
{
    ec_status st = EC_ERR_OOM;
    ec_value *m = ec_map();
    if (!m) {
        goto cleanup;
    }
    if ((st = put_kv(m, "code", ec_text(code))) != EC_OK) { goto cleanup; }
    if (message) {
        if ((st = put_kv(m, "message", ec_text(message))) != EC_OK) { goto cleanup; }
    }
    return ec_entity_make_owning("system/protocol/error", m, out);
cleanup:
    ec_value_free(m);
    return st;
}

ec_status ec_empty_params(ec_entity **out)
{
    ec_value *m = ec_map();      /* empty → 0xA0 (N3) */
    if (!m) {
        return EC_ERR_OOM;
    }
    return ec_entity_make_owning("primitive/any", m, out);
}

ec_status ec_resource_target(const char *target, ec_value **out)
{
    ec_value *m = ec_map();
    if (!m) {
        return EC_ERR_OOM;
    }
    ec_value *k = ec_text("targets");
    ec_value *arr = ec_array();
    ec_value *t = ec_text(target);
    if (!k || !arr || !t || ec_array_push(arr, t) != EC_OK) {
        ec_value_free(k); ec_value_free(arr); ec_value_free(t); ec_value_free(m);
        return EC_ERR_OOM;
    }
    if (ec_map_put(m, k, arr) != EC_OK) {
        ec_value_free(k); ec_value_free(arr); ec_value_free(m);
        return EC_ERR_OOM;
    }
    *out = m;
    return EC_OK;
}

File: protocol-generator/nim/src/wire.nim
## The two L2 wire message builders (§3.2 EXECUTE, §3.3 EXECUTE_RESPONSE) plus the
## small result entities. There are ONLY two root message types on the wire
## (§3.3): `system/protocol/execute` and `system/protocol/execute/response`.
## `hello`/`authenticate` are OPERATIONS on `system/protocol/connect`, not message
## types (§4.1).
##
## Frame := [4-byte big-endian length][ECF-encoded envelope] (§1.6) — the framing
## itself lives in transport.nim.
##
## SPDX-License-Identifier: Apache-2.0

import std/options
import ./ecf
import ./model

const
  ExecuteType* = "system/protocol/execute"
  ResponseType* = "system/protocol/execute/response"
  ConnectUri* = "system/protocol/connect"

proc emptyParams*(): Entity =
  ## Canonical empty params (§3.2): a primitive/any whose data is the empty map.
  makeEntity("primitive/any", mapV(@[]))

proc makeResponse*(requestId: string; status: uint64; resultEntity: Entity): Entity =
  ## EXECUTE_RESPONSE (§3.3): data {request_id, status, result}.
  makeEntity(ResponseType, mapV(@[
    EcPair(key: textV("request_id"), val: textV(requestId)),
    EcPair(key: textV("status"), val: uintV(status)),
    EcPair(key: textV("result"), val: resultEntity.toValue()),
  ]))

proc makeExecute*(requestId, uri, operation: string; params: Entity;
                  author = none(seq[byte]); capability = none(seq[byte]);
                  resource = EcValue(nil)): Entity =
  ## EXECUTE (§3.2). `author`/`capability` are absent on the pre-authorized
  ## `system/protocol/connect` path (§4.2); `resource` is optional.
  var pairs = @[
    EcPair(key: textV("request_id"), val: textV(requestId)),
    EcPair(key: textV("uri"), val: textV(uri)),
    EcPair(key: textV("operation"), val: textV(operation)),
    EcPair(key: textV("params"), val: params.toValue()),
  ]
  if author.isSome: pairs.add EcPair(key: textV("author"), val: bytesV(author.get))
  if capability.isSome: pairs.add EcPair(key: textV("capability"), val: bytesV(capability.get))
  if resource != nil: pairs.add EcPair(key: textV("resource"), val: resource)
  makeEntity(ExecuteType, mapV(pairs))

proc errorResult*(code: string; message = none(string)): Entity =
  ## system/protocol/error result entity (§3.3): {code, message?}.
  var pairs = @[EcPair(key: textV("code"), val: textV(code))]
  if message.isSome: pairs.add EcPair(key: textV("message"), val: textV(message.get))
  makeEntity("system/protocol/error", mapV(pairs))
